### PR TITLE
Addition of API calls checkinvoice and findroute

### PR DESCRIPTION
**Pull request by n1bor**
checkinvoice parses an invoice/PaymentRequest and returns the JSON object. Usecase is that you need to be 100% sure that Eclair is paying the amount you expect! Easy to be factor of 1000 out if someone has a bug!

checkroute takes an invoice and checks if we have a route to pay at that moment. Is useful when you do not want to accept a request for payment you know you can not make.

Includes tests for checkroute.

### CODE REVIEW

**Comment by n1bor** [comment]
Issue fixed - but integration test timed out. Can someone give travis a kick to try again.
We you want a pull that extends the time?

**Comment by n1bor** [comment]
fixed bug in help docs and squashed commits.
And travis worked this time too!

File: eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala [hunk 1]
```diff
@@ -198,6 +198,26 @@ trait Service extends Logging {
...
+                      
+                      case "checkinvoice" => req.params match {
+                        case JString(paymentRequest) :: Nil  => Try(PaymentRequest.read(paymentRequest)) match {
+                          case Success(pr) => {
+                            completeRpc(req.id,pr)
+                          }
+                            // setting the payment amount
+                          case _ => reject(RpcValidationRejection(req.id, s"payment request is not valid"))
+                        }
+                      }
```
Comment (dpad85) [inline eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala]: We should reject the method if the `paymentRequest` argument is missing, with something like:

`case _ => reject(UnknownParamsRejection(req.id, "[payment_request]"))`

File: eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala [hunk 2]
```diff
@@ -198,6 +198,26 @@ trait Service extends Logging {
...
+                      
+                      case "checkroute" => req.params match {
+                        case JString(paymentRequest) :: Nil  => Try(PaymentRequest.read(paymentRequest)) match {
+                          case pr@Success(PaymentRequest(_,Some(amount),_,_,_,_)) => {
+                            completeRpcFuture(req.id, (paymentInitiator ? GetRouteForPayment(amount.amount,pr.get.hash, pr.get.nodeId)).mapTo[PaymentResult])
+                          }
+                          case _ => reject(RpcValidationRejection(req.id, s"payment request is not valid or does not include an amount. Amount needed to check route."))
+                        }
```
Comment (dpad85) [inline eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala]: Same as `checkinvoice`, the method should be rejected if the `paymentRequest` argument is missing

File: eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala [hunk 3]
```diff
@@ -198,6 +198,26 @@ trait Service extends Logging {
...
+                          case _ => reject(RpcValidationRejection(req.id, s"payment request is not valid"))
+                        }
+                      }
+                      
+                      case "checkroute" => req.params match {
+                        case JString(paymentRequest) :: Nil  => Try(PaymentRequest.read(paymentRequest)) match {
+                          case pr@Success(PaymentRequest(_,Some(amount),_,_,_,_)) => {
+                            completeRpcFuture(req.id, (paymentInitiator ? GetRouteForPayment(amount.amount,pr.get.hash, pr.get.nodeId)).mapTo[PaymentResult])
+                          }
+                          case _ => reject(RpcValidationRejection(req.id, s"payment request is not valid or does not include an amount. Amount needed to check route."))
```
Comment (dpad85) [inline eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala]: An amount is not required for a payment request (donation cases) and is not needed to find a route

**Comment by dpad85** [comment]
Great stuff, thanks for the PR!

I noticed that the failures field in the `PaymentFailed` is not propertly serialized to JSON and, when a route can not be found, we get an empty failure:

```json
{
  "paymentHash": "ff75b8778563b85d525569a8259862ee9bc50ea7b5d781015dd957c17eaca496",
  "failures": [
    {
      "t": {}
    }
  ]
}
```

This can be fixed with a ThrowableSerializer, something like:

```scala
class ThrowableSerializer extends CustomSerializer[Throwable](format => ({ null }, {
  case t: Throwable if t.getMessage != null => JString(t.getMessage)
  case t: Throwable => JString(t.getClass.getSimpleName)
}))
```

which then gives

```
...
"failures": [
  {
    "t": "route not found"
  }
```

Could you add it to the `JsonSerializers` file (and to the formats field in `Service`?)

File: eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala [hunk 4]
```diff
@@ -198,6 +198,26 @@ trait Service extends Logging {
                           completeRpcFuture(req.id, (paymentHandler ? ReceivePayment(Some(MilliSatoshi(amountMsat.toLong)), description)).mapTo[PaymentRequest].map(PaymentRequest.write))
                         case _ => reject(UnknownParamsRejection(req.id, "[description] or [amount, description]"))
                       }
+                      
+                      case "checkinvoice" => req.params match {
+                        case JString(paymentRequest) :: Nil  => Try(PaymentRequest.read(paymentRequest)) match {
+                          case Success(pr) => {
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala]: nitpick: remove unnecessary braces around `completeRpc`

File: eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala (hunk 1 above)
Comment (n1bor) [reply eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala]: fixed

File: eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala (hunk 2 above)
Comment (n1bor) [reply eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala]: Fixed

File: eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala (hunk 3 above)
Comment (n1bor) [reply eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala]: Arh - was hoping check route also checked the route had enough funds. Guess it does not.
Fixed

File: eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala (hunk 4 above)
Comment (n1bor) [reply eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala]: Arh - using the router is much simpler! Updated.
Use-case for both these is another app using the API - in checkinvoice case is to check amount is what we are expecting - before paying with send. So zeros fine.

**Comment by dpad85** [comment]
Do you mind if I push 1 or 2 commits on your branch to fix some indentation issues? I also would like to add a serializer for the object returned by the `findroute` call, so that it returns something like:

```js
{
  "hops": "node1 -> node2 -> node3"
  "channels": "shortid1 -> shortid2"
}
```

**Comment by n1bor** [comment]
@dpad85 - sure change however you want. Just want to be able to identify if there is a route, and get from the invoice the amount.
Think the more detail in the API the better. As then future proof for other use-cases.

**Comment by n1bor** [comment]
rebased.. any chance of getting this merged as have some other api changes to come.

File: eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala [hunk 5]
```diff
@@ -221,7 +221,34 @@ trait Service extends Logging {
                           completeRpcFuture(req.id, (paymentHandler ? ReceivePayment(Some(MilliSatoshi(amountMsat.toLong)), description)).mapTo[PaymentRequest].map(PaymentRequest.write))
                         case _ => reject(UnknownParamsRejection(req.id, "[description] or [amount, description]"))
                       }
-                      case "send" => req.params match {
+
+                      case "checkinvoice" => req.params match {
+                        case JString(paymentRequest) :: Nil  => Try(PaymentRequest.read(paymentRequest)) match {
+                          case Success(pr) => {
+                            completeRpc(req.id,pr)
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala]: superfluous { } around `completeRpc(req.id,pr)`

File: eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala [hunk 6]
```diff
@@ -221,7 +221,34 @@ trait Service extends Logging {
...
+                          case Failure(f) => reject(RpcValidationRejection(req.id, s"Payment request is not valid - could not parse: "+f.toString()))
+                        }
+                        case _ => reject(UnknownParamsRejection(req.id, "[payment_request]"))
+                      }
+
+                      case "findroute" => req.params match {
+                        case JString(nodeId) :: Nil if nodeId.length()==66 =>
+                          Try(PublicKey(nodeId)) match {
+                            case Success(pk) =>  completeRpcFuture(req.id, (router ? RouteRequest(appKit.nodeParams.nodeId, pk)  ).mapTo[RouteResponse]  )
+                            case (Failure(_)) => reject(RpcValidationRejection(req.id, s"invalid nodeId hash: '$nodeId'"))
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala]: Remove () around `Failure(_)`

File: eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala [hunk 7]
```diff
@@ -221,7 +221,34 @@ trait Service extends Logging {
...
+                          }
+
+                        case JString(paymentRequest) :: Nil  => Try(PaymentRequest.read(paymentRequest)) match {
+                          case pr@Success(PaymentRequest(_,_,_,_,_,_)) =>
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala]: use `case Success(pr: PaymentRequest) =>` instead

**Comment by n1bor** [comment]
Updated README - sorry about commit message - thought I changed that before I pushed??

**Comment by n1bor** [comment]
annoying timeouts... can we add akka.test.timefactor=10 or something? As expecting to use travis as a performance tester is not a good idea?

**Comment by pm47** [comment]
@n1bor flaky tests were actually due to a regression fixed in 1a8cb2a6947c966c20571054afa4bcf37bc7d603. Believe it or not sometimes tests work as expected! 😁

**Pull request merged**